### PR TITLE
RES-1407 Continuous learning: metrics on full-batch trained and sequentially trained sparse networks

### PR DESCRIPTION
**Pull request by FishAres**
Hopefully fixed contributor issues

### CODE REVIEW

File: projects/continuous_learning/process_gsc_byclass.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+
+import numpy as np
+import torch
+
+from nupic.research.frameworks.pytorch.dataset_utils import PreprocessedDataset
+
+
+def process_gsc_by_class(data_dir=None):
+
+    if data_dir is not None:
```
Comment (subutai) [inline projects/continuous_learning/process_gsc_byclass.py:32]: Why not just:
```
if data_dir is None:
    data_dir = "/home/ec2-user/nta/data/"
```

File: projects/continuous_learning/process_gsc_byclass.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+        data_dir = data_dir
+    else:
+        data_dir = "/home/ec2-user/nta/data/"
+
+    if os.path.isdir(data_dir + "data_classes"):
```
Comment (subutai) [inline projects/continuous_learning/process_gsc_byclass.py:37]:
```
if not os.path.isdir(data_dir + "data_classes"):
    os.mkdir(data_dir + "data_classes")
```

File: projects/continuous_learning/Correlation_experiments.ipynb [hunk 3]
```diff
@@ -0,0 +1,358 @@
...
+#  This program is distributed in the hope that it will be useful,
+#  but WITHOUT ANY WARRANTY; without even the implied warranty of
+#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+#  See the GNU Affero Public License for more details.
+#
+#  You should have received a copy of the GNU Affero Public License
+#  along with this program.  If not, see http://www.gnu.org/licenses.
+#
+#  http://numenta.org/licenses/
+
```
Comment (lscheinkman) [inline projects/continuous_learning/Correlation_experiments.ipynb:19]: Python comments are not allowed in .ipynb files. Put the license in a text or source block not in the raw file

File: nupic/research/frameworks/continuous_learning/cl_utils.py [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+
+import sys
+import time
+
+import torch
+import torch.nn.functional as F
+from tqdm import tqdm
+
+
+def train_multi_model(
```
Comment (lscheinkman) [inline nupic/research/frameworks/continuous_learning/cl_utils.py:30]: This looks like a copy & paste from  `train_model` in `model_utils.y
Why not just add the `freeze_params` to the original `train_model` ?

File: nupic/research/frameworks/continuous_learning/cl_utils.py [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+                                num_images=num_images, time_string=time_string)
+        del loss
+        t0 = time.time()
+
+    if progress_bar is not None:
+        loader.n = loader.total
+        loader.close()
+
+
+def train_model(
```
Comment (lscheinkman) [inline nupic/research/frameworks/continuous_learning/cl_utils.py:143]: The only difference between this function and `train_model` in `model_utils.y` is the `target -= 1`.
Why do you need `target -= 1` on every batch?

Comment (lscheinkman) [reply nupic/research/frameworks/continuous_learning/cl_utils.py:143]: If you need to change the target you should apply that transformation to the dataset. See https://pytorch.org/docs/stable/torchvision/datasets.html

File: nupic/research/frameworks/continuous_learning/cl_utils.py [hunk 6]
```diff
@@ -0,0 +1,256 @@
...
+        del loss
+        t0 = time.time()
+
+    if progress_bar is not None:
+        loader.n = loader.total
+        loader.close()
+
+
+def unravel_index(index, shape):
+    """stole from
```
Comment (lscheinkman) [inline nupic/research/frameworks/continuous_learning/cl_utils.py:249]: `s/stole from/copied from/`

BTW, I cant find this function used anywhere, why do you need it?

File: nupic/research/frameworks/continuous_learning/correlation_metrics.py [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+def register_act(experiment, dp_logs=True, shuffle=False):
+    """ Gets network activations when presented with inputs for each class and runs
+     within- and between-batch pearson correlation and dot products between classes.
+    :param dp_logs (optional): Boolean, determines whether to return
+    the dot product log (base 10)
+    :param shuffle (optional): Boolean, if True will also return the metrics
+     calculated on a shuffled version of the activations
+    """
+
+    layer_list = list(experiment.model.named_children())
```
Comment (lscheinkman) [inline nupic/research/frameworks/continuous_learning/correlation_metrics.py:37]: Why do you need to the convert to `list`?

File: nupic/research/frameworks/continuous_learning/correlation_metrics.py [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+    layer_names = [p[0] for p in layer_list]
+
+    def get_act(name):
+        def hook(model, input_, output):
+            act[name] = output.detach().cpu().numpy()
```
Comment (lscheinkman) [inline nupic/research/frameworks/continuous_learning/correlation_metrics.py:42]: `act` is not defined yet

File: nupic/research/frameworks/continuous_learning/cl_utils.py (hunk 5 above)
Comment (FishAres) [reply nupic/research/frameworks/continuous_learning/cl_utils.py:143]: It was part of me rearranging the train/test dataset by class because pytorch's loss function needs actual digits (as opposed to one-hot vectors) and i wanted my network to accommodate the "zero class" too. I did this so that I wouldn't have to change the full batch training/testing sets for interoperability. If there's a cleaner way to do this I'm game.

File: nupic/research/frameworks/continuous_learning/cl_utils.py (hunk 4 above)
Comment (FishAres) [reply nupic/research/frameworks/continuous_learning/cl_utils.py:30]: Just so people could use the original and not get errors. You recommend changing it?

Comment (lscheinkman) [reply nupic/research/frameworks/continuous_learning/cl_utils.py:30]: Set the `freeze_params` default value to `None` and check before using. We should try to follow the [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself) principle whenever possible

File: nupic/research/frameworks/continuous_learning/cl_utils.py (hunk 5 above)
Comment (lscheinkman) [reply nupic/research/frameworks/continuous_learning/cl_utils.py:143]: A better way to apply transformations to the `target` component of a dataset is via the dataset `target_transform`. See https://pytorch.org/docs/stable/torchvision/datasets.html
What do you mean "zero class" in this context? There are 12 classes in GSC (unknown, silence, zero, one, two, three, four, five, six, seven, eight, nine) and "zero" is one of them.

File: nupic/research/frameworks/continuous_learning/cl_utils.py (hunk 6 above)
Comment (FishAres) [reply nupic/research/frameworks/continuous_learning/cl_utils.py:249]: Oh right, it was for something else but torch.reshape() was more efficient. I'll get rid of it
